File: backend/member-service/src/test/java/com/example/memberservice/config/RedisEmbeddedConfig.java
package com.example.memberservice.config;

import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;
import redis.embedded.RedisServer;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

@Slf4j
@Configuration
public class RedisEmbeddedConfig {

    @Value("${spring.data.redis.host}")
    private String host;

    @Value("${spring.data.redis.port}")
    private int port;

    private RedisServer redisServer;

    @PostConstruct
    private void start() throws IOException {
        int redisPort = isRedisRunning() ? findAvailablePort() : port;

        log.info("Starting redis server on port {}", redisPort);

        redisServer = new RedisServer(redisPort);
        redisServer.start();
    }

    @PreDestroy
    private void stop() throws IOException {
        if (redisServer != null) {
            redisServer.stop();
        }
    }

    private boolean isRedisRunning() throws IOException {
        return isRunning(executeGrepProcessCommand(port));
    }

    public int findAvailablePort() throws IOException {
        for (int port = 10000; port <= 65535; port++) {
            Process process = executeGrepProcessCommand(port);
            if (!isRunning(process)) {
                return port;
            }
        }
        throw new IllegalArgumentException("Not Found Available port: 10000 ~ 65535");
    }

    private Process executeGrepProcessCommand(int port) throws IOException {
        String command = String.format("netstat -nat | grep LISTEN|grep %d", port);
        String[] shell = {"/bin/sh", "-c", command};
        return Runtime.getRuntime().exec(shell);
    }

    private boolean isRunning(Process process) {
        String line;
        StringBuilder pidInfo = new StringBuilder();
        try (BufferedReader input = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
            while ((line = input.readLine()) != null) {
                pidInfo.append(line);
            }
        } catch (Exception e) {
            log.error(e.getMessage());
        }
        return !pidInfo.isEmpty();
    }
}

File: backend/member-service/src/test/kotlin/com/example/memberservice/config/RedisEmbeddedConfig.kt
package com.example.memberservice.config

import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Configuration
import redis.embedded.RedisServer

@Configuration
class RedisEmbeddedConfig(
    @Value("\${spring.data.redis.host}") private val host: String,
    @Value("\${spring.data.redis.port}") private val port: Int,
) {

    private val log = LoggerFactory.getLogger(RedisEmbeddedConfig::class.java)
    private var redisServer: RedisServer? = null

    @PostConstruct
    @Throws(IOException::class)
    fun start() {
        val redisPort = if (isRedisRunning()) findAvailablePort() else port

        log.info("Starting redis server on port {}", redisPort)

        redisServer = RedisServer(redisPort).also { it.start() }
    }

    @PreDestroy
    @Throws(IOException::class)
    fun stop() {
        redisServer?.stop()
    }

    @Throws(IOException::class)
    private fun isRedisRunning(): Boolean = isRunning(executeGrepProcessCommand(port))

    @Throws(IOException::class)
    fun findAvailablePort(): Int {
        for (candidate in 10000..65535) {
            val process = executeGrepProcessCommand(candidate)
            if (!isRunning(process)) {
                return candidate
            }
        }
        throw IllegalArgumentException("Not Found Available port: 10000 ~ 65535")
    }

    @Throws(IOException::class)
    private fun executeGrepProcessCommand(port: Int): Process {
        val command = "netstat -nat | grep LISTEN|grep $port"
        val shell = arrayOf("/bin/sh", "-c", command)
        return Runtime.getRuntime().exec(shell)
    }

    private fun isRunning(process: Process): Boolean {
        val pidInfo = StringBuilder()
        try {
            BufferedReader(InputStreamReader(process.inputStream)).use { input ->
                var line = input.readLine()
                while (line != null) {
                    pidInfo.append(line)
                    line = input.readLine()
                }
            }
        } catch (e: Exception) {
            log.error(e.message)
        }
        return pidInfo.isNotEmpty()
    }
}
